// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd

syntax = "proto3";

package monorail.v3;

option go_package = "api/v3/api_proto";

import "google/protobuf/field_mask.proto";
import "google_proto/google/api/field_behavior.proto";
import "google_proto/google/api/resource.proto";
import "api/v3/api_proto/issue_objects.proto";

// Issues service includes all methods needed for managing Issues.
service Issues {
  // Returns the requested Issue.
  //
  // Raises:
  //   INVALID_ARGUMENT if `name` is formatted incorrectly.
  //   NOT_FOUND if the issue does not exist.
  //   PERMISSION_DENIED if the requester is not allowed to view the issue.
  rpc GetIssue (GetIssueRequest) returns (Issue) {}
  // Returns the requested Issues.
  //
  // Raises:
  //   INVALID_ARGUMENT if `names` is formatted incorrectly or missing.
  //   NOT_FOUND if any of the given issues do not exist.
  //   PERMISSION_DENIED if the requester does not have permission to view one
  //       (or more) of the given issues.
  rpc BatchGetIssues(BatchGetIssuesRequest) returns (BatchGetIssuesResponse) {}
  // Searches over issues within the specified projects.
  // TODO(monorail:6988): Document possible errors when implemented.
  rpc SearchIssues (SearchIssuesRequest) returns (SearchIssuesResponse) {}
  // Lists comments for an issue.
  //
  // Raises:
  //   INVALID_ARGUMENT if `parent` is formatted incorrectly or `page_size` < 0.
  //   NOT_FOUND if `parent` does not exist.
  //   PERMISSION_DENIED if the requester is not allowed to view `parent`.
  rpc ListComments (ListCommentsRequest) returns (ListCommentsResponse) {}
  // Modifies Issues and creates a new Comment for each.
  // We do not offer a standard UpdateIssue because every issue change
  // must result in the side-effect of creating a new issue. We also want
  // to allow for any combination of issue changes to be made at once in
  // a monolithic method.
  //
  // Raises:
  //   INVALID_ARGUMENT required fields are missing or fields are formatted
  //     incorrectly.
  //   NOT_FOUND if any specified issues are not found.
  //   PERMISSION_DENIED if the requester is not allowed to make the
  //     requested change.
  rpc ModifyIssues (ModifyIssuesRequest) returns (ModifyIssuesResponse) {}
  // Makes an issue from an IssueTemplate and deltas.
  //
  // Raises:
  //   TODO(crbug/monorail/7197): Document errors when implemented
  rpc MakeIssueFromTemplate (MakeIssueFromTemplateRequest) returns (Issue) {}
  // Makes a basic issue, does not support phases, approvals, or approval
  // fields.
  // We do not offer a standard CreateIssue because Issue descriptions are
  // required, but not included in the Issue proto.
  //
  // Raises:
  //   TODO(crbug/monorail/7919): Document errors when implemented
  rpc MakeIssue (MakeIssueRequest) returns (Issue) {}
}


// The request message for Issues.GetIssue.
// Next available tag: 2
message GetIssueRequest {
  // The name of the issue to request.
  string name = 1 [
      (google.api.resource_reference) = {type: "api.crbug.com/Issue"},
      (google.api.field_behavior) = REQUIRED ];
}

// The request message for Issues.BatchGetIssues.
// Next available tag: 2
message BatchGetIssuesRequest {
  // The names of the Issues to request.
  repeated string names = 1 [
      (google.api.resource_reference) = {type: "api.crbug.com/Issue"},
      (google.api.field_behavior) = REQUIRED ];
}

// The response message for Issues.BatchGetIssues.
// Next available tag: 2
message BatchGetIssuesResponse {
  // Issues matching the given request.
  repeated Issue issues = 1;
}

// The request message for Issues.SearchIssues.
// Next available tag: 6
message SearchIssuesRequest {
  // The names of Projects in which to search issues.
  // TODO(crbug/monorail/7143): `projects` must hold exactly one project
  // for now. We will add cross-project search later.
  repeated string projects = 1 [
    (google.api.resource_reference) = {type: "api.crbug.com/Project"},
    (google.api.field_behavior) = REQUIRED ];
  // The query string can contain any number of free text and
  // field search expressions.
  // Please see https://bugs.chromium.org/p/chromium/issues/searchtips for more
  // details of how the query string works.
  string query = 2;
  // The maximum number of items to return. The service may return fewer than
  // this value.
  // If unspecified, at most 1000 items will be returned.
  // The maximum value is 1000; values above 1000 will be coerced to 1000.
  int32 page_size = 3;
  // A page token, received from a previous `SearchIssues` call.
  // Provide this to retrieve the subsequent page.
  //
  // When paginating, all other parameters provided to `SearchIssues` must match
  // the call that provided the page token.
  string page_token = 4;
  // The string of comma separated field names used to order the items.
  // Adding '-' before a field, reverses the sort order.
  // E.g. 'stars,-status' sorts the items by number of stars, high to low,
  // then by status, low to high.
  string order_by = 5;
}

// The response message for Issues.SearchIssues.
// Next available tag: 3
message SearchIssuesResponse {
  // Issues matching the given request.
  repeated Issue issues = 1;
  // A token, which can be sent as `page_token` to retrieve the next page.
  // If this field is omitted, there are no subsequent pages.
  string next_page_token = 2;
}

// The request message for Issues.ListComments.
// Next available tag: 4
message ListCommentsRequest {
  // The name of the issue for which to list comments.
  string parent = 1 [
    (google.api.resource_reference) = {type: "api.crbug.com/Issue"},
    (google.api.field_behavior) = REQUIRED ];
  // The maximum number of items to return. The service may return fewer than
  // this value.
  // If unspecified, at most 1000 items will be returned.
  // The maximum value is 1000; values above 1000 will be coerced to 1000.
  int32 page_size = 2;
  // A page token, received from a previous `ListComments` call.
  // Provide this to retrieve the subsequent page.
  //
  // When paginating, all other parameters provided to `ListComments` must
  // match the call that provided the page token.
  string page_token = 3;

  // TODO(monorail:7187): Add filter.
}

// The response message for Issues.ListComments
// Next available tag: 3
message ListCommentsResponse {
  // The comments from the specified issue.
  repeated Comment comments = 1;
  // A token, which can be sent as `page_token` to retrieve the next page.
  // If this field is omitted, there are no subsequent pages.
  string next_page_token = 2;
}


// Holds changes to one issue, used in ModifyIssuesRequest.
// Next available tag: 9
message IssueDelta {
  // The issue's `name` field is used to identify the issue to be
  // updated. `issue.name` must always be filled.
  Issue issue = 1 [
      (google.api.resource_reference) = {type: "api.crbug.com/Issue"},
      (google.api.field_behavior) = REQUIRED ];
  // The list of fields in `issue` to be updated.
  // Repeated fields will be appended to.
  google.protobuf.FieldMask update_mask = 2;

  // Values to remove from the repeated fields of the issue.
  // Values with rule-based Derivation will be ignored.

  // Cc's to remove.
  repeated string ccs_remove = 3 [
      (google.api.resource_reference) = {type: "api.crbug.com/User"}];
  // Blocked_on issues to remove.
  repeated IssueRef blocked_on_issues_remove = 4;
  // Blocking issues to remove.
  repeated IssueRef blocking_issues_remove = 5;
  // Components to remove.
  repeated string components_remove = 6 [
      (google.api.resource_reference) = {type: "api.crbug.com/ComponentDef"}];
  // Labels to remove.
  repeated string labels_remove = 7;
  // FieldValues to remove.
  repeated FieldValue field_vals_remove = 8;

  // TODO(crbug.com/monorail/7657): add Attachment uploading and removing.
}

// The differences to an approval field. Used to update or make issue.
// Next available tag: 6
message ApprovalDelta {
  // The ApprovalValue we want to update.
  ApprovalValue approval_value = 1;
  // Repeated fields found in `update_mask` will be appended to.
  google.protobuf.FieldMask update_mask = 2;
  // Resource names of the approvers we want to remove.
  repeated string approvers_remove = 3 [
      (google.api.resource_reference) = { type: "api.crbug.com/User" }
  ];
  // FieldValues that do not belong to `approval_value` will be ignored.
  repeated FieldValue field_vals_add = 4;
  // FieldValues that do not belong to `approval_value` will be ignored.
  repeated FieldValue field_vals_remove = 5;
  // TODO(crbug.com/monorail/7657): add Attachment uploading and removing.
}


// The type of notification a change should trigger.
// See monorail/doc/userguide/email.md
// Next available tag: 2
enum NotifyType {
  // The default value. This value is unused.
  NOTIFY_TYPE_UNSPECIFIED = 0;
  // An email notification should be sent.
  EMAIL = 1;
  // No notifcation should be triggered at all.
  NO_NOTIFICATION = 2;
}


// The request message for Issues.ModifyIssues.
// Next available tag: 4
message ModifyIssuesRequest {
  // The issue changes to make.
  repeated IssueDelta deltas = 1 [
      (google.api.field_behavior) = REQUIRED ];
  // The type of notification the modifications should trigger.
  NotifyType notify_type = 2;
  // The comment text that should be added to each issue in delta.
  string comment_content = 3;
}


// The response message for Issues.ModifyIssues.
// Next available tag: 2
message ModifyIssuesResponse {
  // The updated issues.
  repeated Issue issues = 1;
}

// The request message for MakeIssueFromTemplate.
// Next available tag: 5
message MakeIssueFromTemplateRequest {
  // Resource name of the template to use for filling in default values
  // and adding approvals and phases.
  string template = 1 [
      (google.api.resource_reference) = { type: "api.crbug.com/Template" }
  ];
  // The issue differences relative to the `template.issue` default.
  IssueDelta template_issue_delta = 2;
  // Changes to fields belonging to approvals relative to template default.
  // While ApprovalDelta can hold additional information, this method only
  // allows adding and removing field values, all other deltas will be ignored.
  repeated ApprovalDelta template_approval_deltas = 3;
  // The issue description, will be saved as the first comment.
  string description = 4;
}

// The request message for MakeIssue.
// Next available tag: 5
message MakeIssueRequest {
  // The name of the project the issue should belong to.
  string parent = 1 [
    (google.api.resource_reference) = {type: "api.crbug.com/Project"},
    (google.api.field_behavior) = REQUIRED ];
  // The issue to be created.
  Issue issue = 2;
  // The issue description.
  string description = 3;
  // The type of notification the creation should trigger.
  NotifyType notify_type = 4;
}
